ARCs/arc-0024.md

```diff
@@ -0,0 +1,485 @@
+application call transaction and the other is by reading global or local state.
+The state variables are named.
+
+An imaginary Application called SugarSupplierContract offers
+a price for refined sugar to the public. This can be read from the
+global state variable "SugarPrice." A retailer negotiates a discount
+wholesale price. This is offered to the customer via a local state variable
+called "YourPrice." The Application also offers a method to calculate 
+the estimated shipping price for a given volume of sugar.
+
+The retailer would like to be integrate with this SugarSupplierContract
+using the development tool of their choice.
+
+The retailer, in this scenario already has the Application Specification
+available, so they give this to their IDE and the following proxy class is
+generated. The below is just a C# example.
+
+```cs
+   public abstract class SugarSupplierContract : SmartContractReference
+    {
+        [Storage(StorageType.Global)]
+        public ulong SugarPrice;
+
+        [Storage(StorageType.Local)]
+        public ulong YourPrice;
+
+  
+        public abstract ulong ShippingCost(ulong VolumeInKilos);
+    } 
+```
+
+Proxies can support server or dapp (frontend) calls to an Algorand node
+or even contract to contract invocations for compilers.
+
+Even without the IDE to tooling support, a dapp author
+still needs to understand how the SugarSupplierContract is intended
+to be used, so information must be provided somehow.
+
+To achieve this goal, this ARC document proposes that the
+Application Specification include a single ARC-4 Contract
+description, and that the ARC-4 Contract section be updated to
+include locals and globals:
+
+```typescript
+interface Contract {
+  /** A user-friendly name for the contract */
+  name: string;
+  /** Optional, user-friendly description for the interface */
+  desc?: string;
+  /**
+   * Optional object listing the contract instances across different networks
+   */
+  networks?: {
+    /**
+     * The key is the base64 genesis hash of the network, and the value contains
+     * information about the deployed contract in the network indicated by the
+     * key
+     */
+    [network: string]: {
+      /** The app ID of the deployed contract in this network */
+      appID: number;
+    }
+  }
+  /** All of the methods that the contract implements */
+  methods: Method[];
+
+  /** All local variables intended for consumption as part of this contract **/
+  locals: Variable[];
+
+  /** All global variables intended for consumption as part of this contract **/
+  globals: Variable[];
+
+}
+```
+
+There needs to be a _Variable_ specification too, following similar style as
+the _Method_ entity:
+
+```typescript
+interface Variable {
+  /** The name of the variable **/
+  name: string;
+  /** Optional, user-friendly description for the variable **/
+  desc?: string;
+  /** The type of the variable **/
+  type: string;
+}
```

Comment:
Are we going to allow ABI types for `Variable`? E.g.: I can specify `bool` so that a client can properly read the App state interpreting a `uint64` as a `bool`.

Reply:
Yes this ARC is quite old now. Since it was written there are a few things I've learned. Yes probably the variable type should be from an agreed list, in ARC4. Though right now arc4's types are far from acceptable.

Reply:
Also the Globals/Locals needs to change as Jannotti suggests above, that global/local not be explicit - as we will have boxes, etc.

Reply:
Yes, the _current_ state schema can not be known on App creation due to Boxes, but I still think would be useful to have a standard way to declare the state schema needed on App create in addition to methods' ABI specs. For example: imagine a client wants to "install" a TEAL dependency, deploy the App and call it. ABI Interface JSON is useful to encode the calls, but the client should still know the initial state schema to deploy the App correctly.